File: html_to_pptx.mjs
/**
 * Convert poster.html → poster.pptx
 * 1. Screenshot HTML at 1920×1080 via puppeteer
 * 2. Embed the image as a full-bleed slide via pptxgenjs
 */
import puppeteer from "puppeteer";
import PptxGenJS from "pptxgenjs";
import { readFileSync } from "fs";
import { fileURLToPath } from "url";
import path from "path";
import { writeFileSync } from "fs";

const __dir  = path.dirname(fileURLToPath(import.meta.url));
const HTML   = path.join(__dir, "poster.html");
const SHOT   = path.join(__dir, "_poster_shot.png");
const OUT    = path.join(__dir, "poster.pptx");
const W = 1920, H = 1080;

// ── 1. Screenshot ──────────────────────────────────────────────────────────
console.log("📸  Screenshotting poster.html …");
const browser = await puppeteer.launch({
  headless: "new",
  args: ["--no-sandbox", "--disable-setuid-sandbox"],
});
const page = await browser.newPage();
await page.setViewport({ width: W, height: H, deviceScaleFactor: 2 }); // 2× for crisp text
await page.goto(`file://${HTML}`, { waitUntil: "networkidle0" });
await new Promise(r => setTimeout(r, 1500)); // let Google Fonts load
const imgBuf = await page.screenshot({ type: "png", fullPage: false });
await browser.close();
writeFileSync(SHOT, imgBuf);
console.log(`   screenshot: ${(imgBuf.length / 1024).toFixed(0)} KB`);

// ── 2. Build PPTX ──────────────────────────────────────────────────────────
console.log("📊  Building poster.pptx …");
const pptx = new PptxGenJS();
pptx.defineLayout({ name: "WIDE", width: 13.33, height: 7.5 });
pptx.layout = "WIDE";

const slide = pptx.addSlide();
const b64   = readFileSync(SHOT).toString("base64");
slide.addImage({
  data: `data:image/png;base64,${b64}`,
  x: 0, y: 0, w: "100%", h: "100%",
});

await pptx.writeFile({ fileName: OUT });
console.log(`✅  Saved → ${OUT}`);

File: html_to_pptx.py
"""
Convert poster.html → poster.pptx
Screenshot the page at 1920×1080 then embed as a full-slide image.
"""
import asyncio, pathlib, io
from playwright.async_api import async_playwright
from pptx import Presentation
from pptx.util import Emu

HTML = pathlib.Path(__file__).parent / "poster.html"
OUT  = pathlib.Path(__file__).parent / "poster.pptx"

W_PX, H_PX = 1920, 1080          # target screenshot resolution
W_EMU = 12192000                  # 1920 × 6350 EMU/px  (13.333 in widescreen)
H_EMU = 6858000                   # 1080 × 6350 EMU/px  (7.5 in widescreen)


async def screenshot() -> bytes:
    async with async_playwright() as p:
        browser = await p.chromium.launch()
        page    = await browser.new_page(viewport={"width": W_PX, "height": H_PX})
        await page.goto(HTML.as_uri(), wait_until="networkidle")
        await page.wait_for_timeout(1500)          # let Google Fonts render
        img = await page.screenshot(full_page=False, type="png")
        await browser.close()
        return img


def build_pptx(img_bytes: bytes) -> None:
    prs = Presentation()
    prs.slide_width  = Emu(W_EMU)
    prs.slide_height = Emu(H_EMU)

    layout = prs.slide_layouts[6]   # blank
    slide  = prs.slides.add_slide(layout)

    slide.shapes.add_picture(
        io.BytesIO(img_bytes),
        left=Emu(0), top=Emu(0),
        width=Emu(W_EMU), height=Emu(H_EMU),
    )
    prs.save(OUT)
    print(f"Saved → {OUT}")


async def main():
    print("Screenshotting poster.html …")
    img = await screenshot()
    print(f"Screenshot: {len(img)//1024} KB")
    build_pptx(img)


asyncio.run(main())
